# Writing Functions


Remember to type in and run
all the examples as you go through these materials. 


## Functions

Another crucial part of programming is the
ability to define functions when we want to do the same thing in several places.  
A function is a rule that takes input values and produces output values.  
Python functions are not quite the same as
mathematical functions.  
However in Python a
function that is called several times with the same input can return
different values each time. A Python function can also have side
effects, such as output to the screen or to a file. 

We will start of by exploring some of the functions that are available
in Python and then we will see how to build our own functions.

### Built-in Functions 

We have already come across several functions that are available
directly in Python: `print`, `int`, `float`, `round`, `list`, `range`...  
Some others are `type`, `abs`, `min` and `max`. 

* To use a function you need the name of the function followed by the arguments in brackets. 
    * The arguments can be numbers, variables, calculations or other functions. 

The functions we had already seen and also the `abs` and `type` function take one input, known as its ***"argument"*** and **return** one thing back, known as the result. 

In [1]:
int(4.436e2), round(4.436e2), float(-68), type(42), type("Hello")

(443, 444, -68.0, int, str)

In [2]:
abs(5), abs(-5), abs(6.01 + 0.01), abs(-6.01 - 0.01)

(5, 5, 6.02, 6.02)

The `min` and `max` functions take two (or more) arguments and return one result. 

If there are several arguments they have to be separated by commas. 

In [3]:
max(-3, 5)

5

In [4]:
min(-3, 5, 0, -10)

-10

## Returned values

Some functions like `sqrt()` **return** a value that can be used, others just perform a function but do not give anything back, such as the `print()` function, which only prints to the screen but does not return a value: 

In [5]:
from math import sqrt
a = sqrt(2)
print(a)

1.4142135623730951


In [6]:
b = print(42)

42


In [7]:
print(b)

None


* A special value of `None` was returned to the varable `b`.

Another very useful built-in function is the help function.  
The help
function takes a single argument and 
prints useful information about the
thing passed on as the input argument.  
We can use this get information about
what a function does and how to use it.

In [ ]:
help(abs)

### Imported Functions

In the previous section we saw that to access certain objects (constants or functions),
we have to load a library first. 

In [ ]:
import math

print(math.exp(1))

In [ ]:
from math import sqrt, cos, sin, pi

print(sqrt(36))
print(cos(0))
print(cos(pi))

In [ ]:
a = 2.3

print(sin(2*a), 2 * sin(a) * cos(a))

In [ ]:
x = 2
print(math.cosh(x), (math.exp(x) + math.exp(-x))/2.0)
print(math.log(math.exp(x)), math.exp(math.log(x)))

The mathematical functions in the `math` module always work with
floating point numbers. This means that integers will be converted to
floating point numbers first. The expression `math.sqrt(36)` will return
the floating point number `6.0` even though the result could be
represented by an integer.

The `help` function can also be applied to modules. This will result in
a list of all the functions and constants contained in that module,
together with their documentation. This information can also be found in
the Python documentation, but it is often convenient to have it
available directly. 

* Try this for the `math` module by printing `help(math)`.

The built-in function `dir` allows us to get just a list of names of
functions and constants for a module.

In [ ]:
import math

print(dir(math))

## Defining Functions 

We can define our own functions using the `def` keyword.

In [ ]:
def sq(x):
    return x**2

* A function definition starts with the keyword `def`, 
* then comes the name of the function and a set of brackets, 
    * The names of functions and parameters have to follow the same rules as
the names of variables  
(letters, numbers and underscores, no numbers or
underscores at the start).
* inside the brackets put any parameters that are to be used in the function
    * multiple parameters are separated by commas.
* a colon must follow the brackets
* After the colon, starting on the next line, we can have one or more lines of code that make up the function.
    * anything the function does must be *indented* by the same number of spaces (conventionally 4)
    * the contents of the function are known as a ***"code block"*** (see below).
    * Anything indented will be run together when the function is used, but nothing afterwards.
    
Once a function is defined, it can be called in exactly the same way as
a built-in function.  
The **`return`** line says what the output of the function will be.  
* The **returned** output value can then be used as a variable or printed.

In [8]:
def sq(x):
    return x**2


print(sq(2))

4


In [9]:
u = 3.5
v = sq(u+0.5) - 1.0
print(v)

15.0


### Exercise: 

Define the function $f(x) = \dfrac{x^2}{3 + x}$. 

* Determine the quantities $f(0)$, $f(-1)$ and $f(1/3)$.

In [16]:
#define function name and arguments here
    #indented code for what to return to outside programme
    
def topolino(x) :
    return (x**2)/(3+x)
print(topolino(0), topolino(-1))


0.0 0.5


Result: `0.0 0.5 0.03333333333333333`

Double click for Solution: $\bullet\bullet$

<!--#model code:
def f(x):
    return x**2/(3 + x)
#tests:
print(f(0), f(-1), f(1/3))
-->

### Indentation of Code Blocks

The code run by the fuction must be *indented* by the same number of spaces.  
This tells Python what is part of the function and what is outside it.  
We can run many commands inside a function, including printing to the screen, rather than returning a value to be printed or used in a variable.

In [17]:
def do_things(a_number):
    print("Your number is %.2f to 2DP" % a_number)
    print("Squared it is:", a_number**2)
    asqrt = a_number**0.5
    print("The square root of " + str(a_number) + " has been returned")
    return asqrt
    print("This does nothing because a return statement terminates the function!")

print("This is not inside the function!")

This is not inside the function!


In [18]:
x = 5.499025
y = do_things(x)
print("This is after the function finishes!")

Your number is 5.50 to 2DP
Squared it is: 30.239275950624997
The square root of 5.499025 has been returned
This is after the function finishes!


* Notice: the last `print()` in the `do_things()` function is ignored because a function **ends** on a `return` statement!

In [ ]:
print(y)

Note: When a function is used any arguments needed must be given in the brackets.  
Also the name of the argument in the function definition `a_number` and that given as an argument when applying the function `x` are **unrelated**.

### Exercise:  Truncate Decimal Places

Write a function `trunc5dp` that truncates a number after the fifth decimal place. For example, `trunc5dp(4321.1234567) = 4321.12345`. 

The steps (algorithm) for doing this are:
1. Multiply by $10^5$ (either written `100000` or `1e5`), - e.g. $4321.1234567$ becomes $432112345.67$
2. use the `int()` function to remove trailing decimals, - e.g. $432112345.67$ becomes $432112345$
3. divide by $10^5$. - e.g. $432112345$ becomes $4321.12345$

In [22]:
#function name definition, with 1 argument
    # code block
    # using algorithm
    # goes here
    # return the answer
    
def trunc5dp (x) :
    x = int(x*(10**5))/(10**5)
    return x

example = trunc5dp(4321.1234567)

print(example)
    
    

    

ans = trunc5dp(4321.1234567)
print(ans)

4321.12345
4321.12345


* Does your function do what you expect for negative numbers?

In [ ]:
print(trunc5dp(-4321.1234567))  # truncates (round towards zero)

Click to see solution...

<!--#code:
def trunc5dp(x):
    x = x*1e5
    x = int(x)
    x = x/1e5
    return x
-->

The parameters inside the definition of the function will be assigned
as the names of the arguments only temporarily inside the function.  
They do not keep these variable names outside.

In [23]:
def myfunc(temporary_variable_name):
    return temporary_variable_name * 2


x = 10

# x will be renamed temporary_variable_name inside the function (and only there)
y = myfunc(x)

print(y)

20


In [24]:
print(temporary_variable_name)  # this will lead to an error

NameError: name 'temporary_variable_name' is not defined

* Think about this...

In the same way you can use a variable name in the definition of a function that is the same as in the rest of the program, and it will not affect it.

In [25]:
a = 1


def root(a):
    a = a**0.5
    return a


b = 2

print(root(b))
print(a)

1.4142135623730951
1


* Here `a` is a different thing *inside* to *outside* the function. Make sure you are clear on this.

Values defined above can be used in the function, but will not be changed.

### Exercise: Calculating Perimeter and Area  (1)

Complete the following programs by writing functions to calculate: 
1. The area of a circle, given its radius ($a = \pi r ^2$);

In [26]:
from math import pi

## Instructions: uncomment the following lines and make the necessary edits

# def circle_area(r):
#     return ??
def circle_area (x) :
    A = pi*(x**2)
    return A


print(circle_area(2.0))

12.566370614359172


Result: `12.566370614359172`

Solution: (double click)

<!--
def circle_area(r):
    return pi * r**2
-->

### Example: No Parameters

In [27]:
def the_answer():
    return 42


ans = the_answer()
print(ans)

42


### Example: Multiple Parameters

In [28]:
def quadratic_value(a, b, c, x):
    "Evaluate a quadratic polynomial a*x**2 + b*x + c"
    return a*x**2 + b*x + c


help(quadratic_value)

Help on function quadratic_value in module __main__:

quadratic_value(a, b, c, x)
    Evaluate a quadratic polynomial a*x**2 + b*x + c



In [29]:
print(quadratic_value(2.0, 3.0, 4.0, 0.5))

6.0


If a function requires several parameters, we need to separate the
parameter names by commas. 

* The example also shows the use of a *documentation string*, which can be useful for the user.
* The `help` function will  print  the documentation string, as well as the name of the function and the
function parameters.

### Local Variables

Any variables that are assigned inside a function definition are called *local*
variables, and **only exist *inside* the function**.  
This means that any local variables that we use for intermediate calculations inside a function will not interfere
with the variables in the rest of the program.

In [13]:
import math


def ellipse(a, b):
    ab = a*b
    print(float(a*b))
    return math.pi*ab
print(ellipse(2,4))

8.0
25.132741228718345


In [11]:
ab = 4  # unrelated to local variable ab in ellipse
print(ab)

4


In [14]:
a = 2.0
b = 3.0

A = ellipse(a, b)
print(ellipse(2,3))

6.0
6.0
18.84955592153876


* Note: `ellipse()` *printed* the **internal** value of `ab` to the output, but ***returned*** te value of $\pi a b$ to the variable `A`:

In [ ]:
print(A)

#the value of ab defined outside the function remains unchanged!
print(ab)

### Multiple Return Values 

A function can return several values, by separating
the values by commas in the return statement. 

The following example shows a function that returns the two roots of the quadratic equation
$a x^2 + b x + c = 0$. 

In [15]:
import math


def quadratic(a, b, c):
    "return the roots of the quadratic equation a*x**2 + b*x + c = 0."
    discriminant = b**2 - 4*a*c  # assumed to be positive
    offset = math.sqrt(discriminant)
    x1 = (-b + offset)/(2*a)
    x2 = (-b - offset)/(2*a)
    return x1, x2


print(quadratic(4.0, 0.0, -16.0))

(2.0, -2.0)


We can assign names to each of the return values using the following
syntax:

In [16]:
root1, root2 = quadratic(4.0, 0.0, -16.0)
print(root1)
print(root2)

2.0
-2.0


* Note that the names of the local variables (`x1` and `x2`) do not have
to be the same as the variables we use to name the results of the
function (`root1`, `root2`).

### Exercise: Calculating Perimeter and Area  (2)

2. the perimeter and area of a square, given the side length ($P = 4L$, $A=L^2$);

In [24]:
## Instructions: uncomment the following lines and make the necessary edits

# def ??:
#     return ??, ??

def square_geometry(l) :
    P = 4*l
    A = l**2
    print(P,A)
    return P,A
    

print(square_geometry(3))

12 9
(12, 9)


Result: `(12, 9)`

<!--
def square_geometry(s):
    return 4*s, s**2
-->

### Exercise: Calculating Perimeter and Area  (3)

3. the perimeter and area of a rectangle, given the width and height ($P=2W + 2H$, $A=W\times H$). 

In [28]:
## Instructions: uncomment the following lines and make the necessary edits

# ?? ??(??, ??):
#     ??
def rectangle_geometry (W,H) :
    P = (2*W)+(2*H)
    A = W*H
    print(P,A)
    return (P,A)
P,A = rectangle_geometry(3, 4)

print (P,A)

14 12
14 12


### No Return Values

It is possible to define functions that do not return any values at all.
These are used to perform tasks such as printing, reading or writing files, accessing the internet, or plotting graphs.

In [29]:
def hello():
    print("Hello World!")


hello()

Hello World!


The function does not contain a `return` statement and
therefore the function does not return a value, just a special value
`None`.

In [30]:
greeting = hello()  # the action of printing will still be performed

print(greeting)

Hello World!
None


### Optional Arguments and Keyword Arguments 

Sometimes arguments can be given default values using an equals sign. 

If that argument is not entered when applying the function it takes the default value.

In [31]:
def raisepower(base, exponent=2):
    return base**exponent


print(raisepower(10))
print(raisepower(10, 5))
print(raisepower(2, exponent=10))

100
100000
1024


* Keyword arguments must appear **after** compulsory arguments that do not have default values. 
* If there are more than 3 or 4 arguments it is a good idea to have them as optional keyword arguments with default values.

### Functions as Values 

Functions can also be used as input *values* in Python. 

In [48]:
from math import cos, sin, tan, pi


def f_zero_ten(InputFunction):
    print(InputFunction(0.0), InputFunction(10.0), InputFunction(2*pi))


f_zero_ten(cos)
f_zero_ten(sin)
f_zero_ten(tan)

1.0 -0.8390715290764524 1.0
0.0 -0.5440211108893698 -2.4492935982947064e-16
0.0 0.6483608274590866 -2.4492935982947064e-16


In [43]:
def sq(x):
    return x**2


f_zero_ten(sq)

0.0 100.0 39.47841760435743


We can also use an assignment (**`=`**) statement to assign a name to a function.

In [49]:
import math

f = math.cos
print(f(0.1))

0.9950041652780258


### Exercises

* Define a function `two_times_x()` that takes a single argument `x` and *returns* the value $2x$.

In [52]:
def two_times_x (x):
    x = 2*x
    return x

a1 = two_times_x(10)
print(a1)

20


Result: `20`

Solution (click):  

<!--
def two_times_x(x):
    return 2*x
-->

* Define a function `twice()` that takes a function `f` and a value `x` as two arguments, then returns $f(f(x))$

In [54]:
def twice(f, x) :
    return f(f(x))

a2 = twice(two_times_x, 10)
print(a2)

40


Result: `40`

Solution (click):

<!--
def twice(f, x):
    return f(f(x))
-->

* Define a function `thrice()` that takes a function `f` and a value `x` as two arguments, then returns $f(f(f(x)))$

In [55]:
def thrice (f,x) :
    return f(f(f(x)))


a3 = thrice(two_times_x, 10)
print(a3)

80


Result: `80`

Solution:

<!--
def thrice(f, x):
    return f(f(f(x)))
-->

## Summary

In [ ]:
# importing a module
import math

# defining a function
def sumsquares(a, b):
    return a**2 + b**2

x = 3.0
y = 2.0
z = sumsquares(x, 2*y)  # calling a function

h = math.sqrt(z)  # calling a built-in function
print(z, h)

## Pitfalls


Watch out for the following sources of errors:

-   Missing colon before indentation block.

-   Wrong or inconsistent indentation. Use for 4 spaces (not tabs) per
    indentation level.

## Task 4: 

## Numerical Differentiation

#### The tasks here lead on from each other. 

#### Submit only PART 2 (steps 1,2,3) in one SCRIPT.py file (ensure you tested it outside the notebook before submission)

### Background

Differentiation of a function $f(x)$ about any point $x=a$ can be defined as follows:
$$f'(a) = \lim\limits_{\delta x \rightarrow 0} \frac{f(a + \delta x) - f(a)}{\delta x}$$

Therefore a numerical approximation to the gradient at a point $a$ can be obtained by evaluating `f(a)`and `f(a+dx)` after
some small increment $\delta x$ (which we name as `dx` for simplicity)

The code below uses this technique to differentiate the specific function for the area of a cube:  
$A=6 L^2,$  
with side length $L=2$

In [56]:
L = 2
dx = 0.1

r1 = 6.0 * L**2.0
r2 = 6.0 * (L+dx)**2.0
f_x = (r2 - r1)/dx

print(f_x)

24.60000000000001


* Look back at `"floating point precision"` in week 1 to explain the last digit.


## PART 1 (warm-up exercise, do not submit this)
* Copy the code above and change change the value of $\delta x$ to 0.01 to see how it affects the numerical approximation to the true value of 24.

In [57]:
L = 2
dx = 0.01

r1 = 6.0 * L**2.0
r2 = 6.0 * (L+dx)**2.0
f_x = (r2 - r1)/dx

print(f_x)

24.05999999999935


* Define a Python function `deriv` that returns the derivative of $6L^2$ at any point `L` using a step size `dx` (use `(L, dx)` as two arguments to the function).

In [58]:
def deriv(L, dx) :
    r1 = 6.0 * L**2.0
    r2 = 6.0 * (L+dx)**2.0
    f_x = (r2 - r1)/dx
    return f_x


print(deriv(2, 0.0001))

24.000600000064765


Result: `24.000600000064765`

## PART 2 (for assessment): A function to take *any* function as an argument and differentiate it

The steps for doing the above procedure for any given function are as follows:
1. Define a Python function `diff(f, a, dx)` to numerically approximate the derivative of *any* function `f(x)` at *any* point `x=a`, using the formula given in part 1;
2. Define the new function you want to differentiate: `g(x)` which *returns* the value of the function `g` around `x`;
3. Give the function name `g`, **without** parentheses `()`, and a new point `b` as arguments to `diff(g, b)`, to return the derivative of `g(x=b)`.  


* **More detail is given below:**

### Step 1
Using the notes to guide you, write a function called `diff()` that takes three arguments: 1. another function *internally named as* `fi` (or whatever), 2. a point `a`, and 3. a small increment `dx`:
* define a new function called differentiate with three arguments: `f`, `a` and `dx`.
* the `diff()` function should evaluate `r1 = f(a)` and `r2 = f(a+dx)` and use these values in the formula for the gradient given in the Background and call it `f_x` (or whatever you like).
* The `diff()` function should `return` the value of `f_x`

### Step 2
Define a new function called `test_function(x)` that returns the result of $g(x) = x^3 - 2x^2 - 5$


### Step 3
Give the test function name as the first argument to your `diff()` function and the values `x=10` and `dx=1e-3` (or `0.001`) as the other two arguments.

In [2]:
# here define a function called diff(internal_function_name, OTHER, ARGUMENTS):
#    r1 = result of applying the internal function named fi at x: i.e. fi(x)
#    and r2 should be fi evaluated at x+dx...
#    result of the differentiation formula gives some value (call it f_x or something)
#    then return f_x


def diff(fi, a, dx) : 
    r1 = fi(a)
    r2 = fi(a+dx)
    f_x = (r2 - r1)/dx
    return f_x



# leave this function as it is here
def test_function(x):
    return (x**3) - (2*(x**2)) - 5


a=10
dx=1e-3 
answer = diff(test_function,a,dx)
#apply your diff function here with arguments of the test function name, the value of x and the step-size.
print(answer) 


260.02800099990964


Result: `260.02800099990964`
* precision may vary on a non-azure machine, don't worry about this if it's OK here.

#### Once it's working put the code above into a python script, check in Spyder (etc.) it works as expected from scratch and submit.

## Extra Practice Example & Building in Loops:

### Iterating a Function 

* Define the function $\displaystyle f(x) = \frac{x+\frac{2}{x}}{2}.$  

* Calculate the values $f(1)$, $f(f(1))$, $f(f(f(1)))$, ... What do you observe?

Click here for a hint...

<!--
Result should be: 1.0 1.5 1.4166666666666665 1.4142156862745097
Try squaring $x_4$ to see what you get!
-->

In [ ]:
# this is equivalent:
x0 = 1.0
x1 = f(x0)
x2 = f(x1)
x3 = f(x2)
print(x0, x1, x2, x3)

* Wrap the function in a loop to apply it N times and experiment to see when it converges to the acccuracy of your output.

In [ ]:
#assign initial x
#put a for loop header here
    #call your function here with x as an argument and return the value back to x

#print the last value of x here
#print the value of x**2 and consider if it's what you expect and why?

Answer: Should converge at $N=5$

Click for solution...

<!--
x=1
for i in range(5):
    x = f(x)
    
print(x)
print(x**2)
-->

* Put the previous code in a function `iterate_func()` that takes three arguments: `f`, `x0` and `N` and returns the final iterated `x`

(click for hint)

<!--
def function_name(ARGUMENTS):
    for loop goes here:
        assign x to new value based on f(x)
    return value of x here
-->

In [ ]:
? ?(?):
    ? ? ? ?(?):
        ? = ?
    ? ?
    
ans = iterate_func(f, 1, 5)
print(ans, ans**2)

Result: `1.414213562373095 1.9999999999999996`

* Finally make a new function `square_root` that takes one argument `(N)` and iterates for 10 recursions the formula $\displaystyle x_{i+1} = \frac{x_{i}+\frac{N}{x_i}}{2}$ 

In [ ]:
#definition of a function 
#including a loop
#applying a formula
#returning a value

print([square_root(val) for val in [4,9,16,25]])

Result: `[2.0, 3.0, 4.0, 5.0]`

Solution (click to reveal):

<!--
def square_root(N):
    x=1
    for i in range(10):
        x = (x + N/x) / 2
    return x
-->